connectors/kafka/src/main/java/forklift/controller/AcknowledgedRecordHandler.java

```diff
@@ -0,0 +1,121 @@
+package forklift.controller;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Maintains a batch of acknowledged records and provides management for adding and removing {@link org.apache.kafka.common.TopicPartition partitions}.
+ * Acknowledged records are those that have started processing but have not yet been committed to the Kafka Broker.  This class is threadsafe.
+ */
+public class AcknowledgedRecordHandler {
+    private static final Logger log = LoggerFactory.getLogger(AcknowledgedRecordHandler.class);
+    private ConcurrentHashMap<TopicPartition, OffsetAndMetadata> pendingOffsets = new ConcurrentHashMap<>();
+    private Set<TopicPartition> assignment = ConcurrentHashMap.newKeySet();
+    private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    /**
+     * Acknowledges that a record has been received before processing begins.  True is returned if processing should occur else false.
+     * Only records belonging to {@link #addPartitions(java.util.Collection) added partitions} may be processed. Note that this is a
+     * blocking method and a short delay may occur should the available topic paritions be changing.
+     *
+     * @param record the record to acknowledge
+     * @return true if the record has been acknowledged and may be processed, else false
+     * @throws InterruptedException if interrupted
+     */
+    public boolean acknowledgeRecord(ConsumerRecord<?, ?> record) throws InterruptedException {
+        lock.readLock().lock();
+        try {
+            boolean acknowledged;
+            TopicPartition topicPartition = new TopicPartition(record.topic(), record.partition());
+            if (!assignment.contains(topicPartition)) {
+                acknowledged = false;
+            } else {
+                long commitOffset = record.offset() + 1;
+                OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(commitOffset, "Commit From Forklift Server");
+                pendingOffsets.merge(topicPartition, offsetAndMetadata, this::greaterOffset);
+                acknowledged = true;
+            }
+            return acknowledged;
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    private OffsetAndMetadata greaterOffset(OffsetAndMetadata a, OffsetAndMetadata b) {
+        if (a == null) {
```

Comment:
You shouldn't _need_ to check for null here, since the merge function will only be called if there is an existing value (which should always be non-null).